### PR TITLE
Stop Auth methods from polling the config on every req.

### PR DESCRIPTION
**Pull request by anoadragon453**
Fixes https://github.com/matrix-org/synapse/issues/7419

Polls all the config options on server setup instead of every req. `check_auth_blocking`, which used the majority of the config options, was moved to a separate class to keep `Auth` clean.

Merging to the release branch to help get this out on matrix.org.

Edit: Lots of tests needed to be updated as they modified the hs' config after the hs had been initialized, which means `check_auth_blocking`, which pulls options on hs intiialization, would be using stale values. These tests were changed such that they modify values in the new `AuthBlocking` class instead of the hs' config.

### CODE REVIEW

File: synapse/api/auth.py [hunk 1]
```diff
@@ -78,6 +78,18 @@ def __init__(self, hs):
...
+        self._server_notices_mxid = self.hs.config.server_notices_mxid
+        self._hs_disabled = self.hs.config.hs_disabled
+        self._limit_usage_by_mau = self.hs.config.limit_usage_by_mau
+        self._hs_disabled_message = self.hs.config.hs_disabled_message
+        self._admin_contact = self.hs.config.admin_contact
+        self._limit_usage_by_mau = self.hs.config.limit_usage_by_mau
+        self._mau_limits_reserved_threepids = (
+            self.hs.config.mau_limits_reserved_threepids
+        )
+        self._max_mau_value = self.hs.config.max_mau_value
```
Comment (richvdh) [inline synapse/api/auth.py:92]: can we factor out a separate class rather than add all these fields to this big class?

Comment (anoadragon453) [reply synapse/api/auth.py:92]: How do you mean? Moving the methods that use these to a new class or putting all these config options in a new inline class?

Comment (richvdh) [reply synapse/api/auth.py:92]: move the methods to a new class. dunno how easy that is

Comment (anoadragon453) [reply synapse/api/auth.py:92]: It's not too bad to move, I just need to update all the references to it.

I could start with just having `Auth.check_auth_blocking` call the new method though.

Or I could just go ahead and update all the references. My plan is:

* Create a new class `AuthBlocking`
* It just handles this method
* Have to add a `build_auth_blocking` method to the hs in addition to `build_auth`.

Before I get too far into any of this, does this sound sane or should we back out?

Comment (richvdh) [reply synapse/api/auth.py:92]: > I could start with just having `Auth.check_auth_blocking` call the new method though.

sounds sensible for now

>     * Create a new class `AuthBlocking`
>     * It just handles this method

yes

>     * Have to add a `build_auth_blocking` method to the hs in addition to `build_auth`.

I'd have `Auth.__init__` build the new object

Comment (anoadragon453) [reply synapse/api/auth.py]: Unfortunately there's quite a few tests that changed the hs's config after the hs had already been created, which need fixing up.

**Pull request merged**